### PR TITLE
Refactor pg_dump queries to gather partition definitions

### PR DESCRIPTION
**Pull request by bmdoil**
Commit 1:
Currently, when backing up a GP5 or GP6 database, getTables will
hold a lock on all partition tables in the database by calling
`pg_get_partition_def` and `pg_get_partition_template_def` on each table,
resulting in tables being locked that are not part of the backup set.
    
This commit extracts the `pg_get_partition_def` and
`pg_get_partition_template_def` function calls out of getTables and
into getPartitionDefs. Now, only partition tables in the backup set  will be
included in calls to these functions.

Commit 2:
Postpone calls of unsafe server-side functions in pg_dump.

Backported from PG15 e3fcbbd

Conflicts resolved:
* OidOptions enum value zeroIsError does not exist, instead used the
  existing zeroAsOpaque.
* Minor conflicts in dumpTableSchema resulting from foreign server
  related code introduced in PG13 commit 4e62091

Original commit message follows:

Avoid calling pg_get_partkeydef(), pg_get_expr(relpartbound),
and regtypeout until we have lock on the relevant tables.
The existing coding is at serious risk of failure if there
are any concurrent DROP TABLE commands going on --- including
drops of other sessions' temp tables.

Arguably this is a bug fix that should be back-patched, but it's
moderately invasive and we've not had all that many complaints
about such failures.  Let's just put it in HEAD for now.

Discussion: https://postgr.es/m/2273648.1634764485@sss.pgh.pa.us
Discussion: https://postgr.es/m/7d7eb6128f40401d81b3b7a898b6b4de@W2012-02.nidsa.loc

For main branch the issue could only occur when backing up from a GP5/GP6 server outside the gpupgrade workflow.
Since src/test/regress is testing main pg_dump against main server, a regress test is not useful (no calls to pg_get_partition_def). We don't have a test framework for backing up older server versions using most recent pg_dump, and that use-case is not supported. [See the 6X PR for the relevant test case](https://github.com/greenplum-db/gpdb/pull/15964/commits/ef9ba02e2df0cb4e9ec75d130e11cc818f889a4b).

### CODE REVIEW

File: src/bin/pg_dump/pg_dump.c [hunk 1]
```diff
@@ -6503,10 +6504,10 @@ getTables(Archive *fout, int *numTables)
 
 	if (fout->remoteVersion >= 90000)
 		appendPQExpBufferStr(query,
-						  "CASE WHEN c.reloftype <> 0 THEN c.reloftype::pg_catalog.regtype ELSE NULL END AS reloftype, ");
+							 "c.reloftype, ");
```
Comment (AJR-VMware) [inline src/bin/pg_dump/pg_dump.c:6507]: nit: whitespace difference

File: src/bin/pg_dump/pg_dump.c [hunk 2]
```diff
@@ -16772,8 +16854,10 @@ dumpTableSchema(Archive *fout, const TableInfo *tbinfo)
 		 * Attach to type, if reloftype; except in case of a binary upgrade,
 		 * we dump the table normally and attach it to the type afterward.
 		 */
-		if (tbinfo->reloftype && !dopt->binary_upgrade)
-			appendPQExpBuffer(q, " OF %s", tbinfo->reloftype);
+		if (OidIsValid(tbinfo->reloftype) && !dopt->binary_upgrade)
```
Comment (AJR-VMware) [inline src/bin/pg_dump/pg_dump.c:16857]: Is this part of the backport?  Seems unrelated.

Comment (bmdoil) [reply src/bin/pg_dump/pg_dump.c:16857]: Yes it's part of the cherry-pick, `avoid calling pg_get_partkeydef(), pg_get_expr(relpartbound),
and regtypeout until we have lock on the relevant tables.` It's the regtypeout aspect of the commit.

**Review by AJR-VMware — APPROVED** [review]
Looks good to me!
